### PR TITLE
Fix FastlanePty.spawn on Windows

### LINKED ISSUE #12144: No child processes (Errno::ECHILD) command_executor.rb:70
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When I run fastlane screengrab from cmd on Windows 10 I got the error below.
The same sourcecode works fine on Mac OS X with the same configuration and emulator which is:
Nexus_5X_API_24_Android_7.0

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>[✔] 🚀
[09:40:31]: fastlane detected a Gemfile in the current directory
[09:40:31]: however it seems like you don't use `bundle exec`
[09:40:31]: to launch fastlane faster, please use
[09:40:31]:
[09:40:31]: $ bundle exec fastlane screengrab
[09:40:31]:
[09:40:31]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
'cat' is not recognized as an internal or external command,
operable program or batch file.
[09:40:31]: Successfully loaded 'C:/Users/username/Documents/Projects/Android/xxx-Android/fastlane/Screengrabfile' 📄

+----------------------------+-------------------------------------------------------------------+
|                        Detected Values from './fastlane/Screengrabfile'                        |
+----------------------------+-------------------------------------------------------------------+
| app_package_name           | com.xxx.xxx                                           |
| app_apk_path               | app/build/outputs/apk/debug/app-debug.apk                         |
| tests_apk_path             | app/build/outputs/apk/androidTest/debug/app-debug-androidTest.apk |
| locales                    | ["en-GB"]                                                         |
| clear_previous_screenshots | true                                                              |
+----------------------------+-------------------------------------------------------------------+

[09:40:31]: The `aapt` command could not be found on your PATH
[09:40:31]: Please ensure that the Android SDK is installed and you have the build tools downloaded and present on your PATH

+-----------------------------+-------------------------------------------------------------------+
|                                  Summary for screengrab 2.86.2                                  |
+-----------------------------+-------------------------------------------------------------------+
| tests_package_name          | com.xxx.xxx.test                                      |
| locales                     | ["en-GB"]                                                         |
| clear_previous_screenshots  | true                                                              |
| output_directory            | fastlane/metadata/android                                         |
| skip_open_summary           | true                                                              |
| app_package_name            | com.xxx.xxx                                           |
| test_instrumentation_runner | android.support.test.runner.AndroidJUnitRunner                    |
| ending_locale               | en-US                                                             |
| app_apk_path                | app/build/outputs/apk/debug/app-debug.apk                         |
| tests_apk_path              | app/build/outputs/apk/androidTest/debug/app-debug-androidTest.apk |
| device_type                 | phone                                                             |
| exit_on_test_failure        | true                                                              |
| reinstall_app               | false                                                             |
+-----------------------------+-------------------------------------------------------------------+

[09:40:32]: Limiting the test classes run by `screengrab` to just those that generate screenshots can make runs faster.
[09:40:32]: Consider using the :use_tests_in_classes or :use_tests_in_packages option, and organize your tests accordingly.
[09:40:32]: Clearing phoneScreenshots within fastlane/metadata/android
[09:40:32]: $ adb devices -l
[09:40:32]: ▸ List of devices attached
[09:40:32]: ▸ emulator-5554          device product:sdk_google_phone_x86 model:Android_SDK_built_for_x86 device:generic_x86 transport_id:1
List of devices attached
emulator-5554          device product:sdk_google_phone_x86 model:Android_SDK_built_for_x86 device:generic_x86 transport_id:1

No child processes

Looking for related GitHub issues on fastlane/fastlane...

➡️  Fastlane doesn't wait on its child processes before returning
    https://github.com/fastlane/fastlane/issues/2092 [closed] 4 💬
    22 Sep 2016

➡️  Fastlane doesn't wait on its child processes before returning
    https://github.com/fastlane/fastlane/issues/2093 [closed] 4 💬
    22 Sep 2016

➡️  Auto-generated README malformation (missing newline before new header)
    https://github.com/fastlane/fastlane/issues/8560 [closed] 4 💬
    03 Aug 2017

and 3 more at: https://github.com/fastlane/fastlane/search?q=No%20child%20processes&type=Issues&utf8=✓

C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/command_executor.rb:70:in `wait': [!] No child processes (Errno::ECHILD)
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/command_executor.rb:70:in `block in execute'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/fastlane_pty.rb:14:in `block in spawn'
        from C:/Ruby24-x64/lib/ruby/2.4.0/open3.rb:205:in `popen_run'
        from C:/Ruby24-x64/lib/ruby/2.4.0/open3.rb:190:in `popen2e'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/fastlane_pty.rb:13:in `rescue in spawn'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/fastlane_pty.rb:6:in `spawn'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/command_executor.rb:51:in `execute'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/screengrab/lib/screengrab/runner.rb:351:in `run_adb_command'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/screengrab/lib/screengrab/runner.rb:80:in `select_device'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/screengrab/lib/screengrab/runner.rb:56:in `run'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/screengrab/lib/screengrab/commands_generator.rb:49:in `block (2 levels) in run'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:75:in `run!'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/screengrab/lib/screengrab/commands_generator.rb:67:in `run'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/screengrab/lib/screengrab/commands_generator.rb:19:in `start'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/fastlane/lib/fastlane/cli_tools_distributor.rb:99:in `take_off'
        from C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/fastlane-2.86.2/bin/fastlane:20:in `<top (required)>'
        from C:/Ruby24-x64/bin/fastlane:23:in `load'
        from C:/Ruby24-x64/bin/fastlane:23:in `<main>'</pre>
</details>

### Environment

<details>
  <pre><details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                        |
| --------------------------- | ---------------------------- |
| OS                          | UNKNOWN                      |
| Ruby                        | 2.4.3                        |
| Bundler?                    | false                        |
| Git                         | git version 2.16.2.windows.1 |
| Installation Source         | C:/Ruby24-x64/bin/fastlane   |
| Host                        | Unknown  ()                  |
| Ruby Lib Dir                | C:/Ruby24-x64/lib            |
| OpenSSL Version             | OpenSSL 1.0.2n  7 Dec 2017   |
| Is contained                | false                        |
| Is homebrew                 | false                        |
| Is installed via Fabric.app | false                        |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# This file contains the fastlane.tools configuration
# You can find the documentation at https://docs.fastlane.tools
#
# For a list of all available actions, check out
#
#     https://docs.fastlane.tools/actions
#

# Uncomment the line if you want fastlane to automatically update itself
# update_fastlane

default_platform(:android)

platform :android do
  desc "Runs all the tests"
  lane :test do
    gradle(task: "test")
  end

  desc "Submit a new Beta Build to Crashlytics Beta"
  lane :beta do
    gradle(task: "clean assembleRelease")
    crashlytics

    # sh "your_script.sh"
    # You can also use other beta testing services here
  end

  desc "Deploy a new version to the Google Play"
  lane :deploy do
    gradle(task: "clean assembleRelease")
    upload_to_play_store
  end
end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
json_key_file("fastlane/key.json")
package_name("com.xxx.xxx")

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.86.2  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.1.0        |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.2        |
| CFPropertyList            | 2.3.6        |
| claide                    | 1.0.2        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.5.6        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.2.8        |
| terminal-notifier         | 1.8.0        |
| unicode-display_width     | 1.3.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.4.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.2        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.6.4        |
| tty-cursor                | 0.5.0        |
| tty-spinner               | 0.8.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| faraday                   | 0.14.0       |
| unf_ext                   | 0.0.7.5      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20170404 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| fastimage                 | 2.1.1        |
| gh_inspector              | 1.1.3        |
| mini_magick               | 4.5.1        |
| multi_json                | 1.13.1       |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.2.1        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.2.1        |
| bundler                   | 1.16.1       |
| faraday_middleware        | 0.12.2       |
| naturally                 | 2.1.0        |
| simctl                    | 1.6.2        |
| uber                      | 0.1.0        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.1        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| little-plugger            | 1.1.4        |
| logging                   | 2.2.2        |
| jwt                       | 2.1.0        |
| memoist                   | 0.16.0       |
| os                        | 0.9.6        |
| signet                    | 0.8.1        |
| googleauth                | 0.6.2        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.13.6       |
| openssl                   | 2.0.5        |
| excon                     | 0.61.0       |
| json                      | 2.1.0        |
| io-console                | 0.4.6        |
</details>

*generated on:* **2018-03-23**
</details></pre>
</details>

### PR DESCRIPTION
**Pull request by janpio**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

As reported in #12144 fastlane currently can't successfully execute commands via `FastlaneCore::CommandExecutor.execute` on Windows (which screengrab does with `adb devices -l` to get a list of devices to run the tests on).

`FastlaneCore::CommandExecutor.execute` internally uses `FastlaneCore::FastlanePty.spawn` to spawn the command, which first tries to `PTY.spawn` and if that fails - which it does on Windows as PTY.spawn is not available - uses an alternative method of `Open3.popen2e`.

Problem is, that the code in `CommandExecutor.execute` tries to `Process.wait(pid)` to fill `$?`, from which it tries to get the exit status via `$?.exitstatus`. On Windows `Process.wait` triggers the `No child processes (Errno::ECHILD)` exception, as documented: https://www.w3cschool.cn/doc_ruby_2_4/ruby_2_4-process.html#method-c-wait

### Description

Lucky for us, the code I cobbled/stole together that is executed on Windows via `Open3.popen2e` already semi takes this into account - it returns the exit status code, making the `Process.wait` and `$?.exitstatus` unnecessary in this case.

I moved both `Process.wait` and `$?.exitstatus` to the `PTY.spawn` part of `FastlaneCore::FastlanePty.spawn` and return the exit status similat ro how the `Open3.popen2e` already works.

There is one test that tests for some EIO exception special case. I had to get rid of 2 `expect`s there - the code was moved inside a function that is not actually run by that test.

To avoid this bug in the future, I also added a test that runs a super simple command via  `CommandExecutor.execute` - without all these changes here it fails, now it doesn't.

---

closes #12144

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: fastlane_core/spec/command_executor_spec.rb [hunk 1]
```diff
@@ -10,11 +15,13 @@
         expect(explodes_on_strip).to receive(:strip).and_raise(Errno::EIO)
 
         child_process_id = 1
-        expect(Process).to receive(:wait).with(child_process_id)
+        unless FastlaneCore::Helper.windows?
+          # expect(Process).to receive(:wait).with(child_process_id) // moved over to fastlane_pty.rb
```
Comment (janpio) [inline fastlane_core/spec/command_executor_spec.rb]: Can I somehow test the calling of `Process.wait()` here in its new place, where I moved it?

Comment (joshdholtz) [reply fastlane_core/spec/command_executor_spec.rb]: Does this same `expect` not work here? I feel like it should 🤔 `Process` should still receive `:wait` - it shouldn't matter that it was moved into `fastlane_pty.rb`

Comment (janpio) [reply fastlane_core/spec/command_executor_spec.rb]: Maybe my logic of the move is wrong then? I couldn't get it to work yesterday :/

Comment (janpio) [reply fastlane_core/spec/command_executor_spec.rb]: Let me test (via CI).

Comment (janpio) [reply fastlane_core/spec/command_executor_spec.rb]: An idea: 

When `Process.wait` was in its old location, it was _always_ executed in this strange tests that tests for some edge case with EIO exception that I don't understand.

Now that I moved it, this doesn't happen any more when this exception is triggered because of the changed location.

Does that make sense?

Comment (janpio) [reply fastlane_core/spec/command_executor_spec.rb]: Ok, I finally understood: This EIO test doesn't _actually_ execute the `.spawn` but "mock" or "fake" it, that's why the `Process.wait` I moved into the code didn't get executed. Will update below.

**Comment by janpio** [comment]
Reworked the code and force pushed - please note of the new PR description. 
(Code now is much nicer as before - I'm happy that I tackled it again)

**Review by joshdholtz — APPROVED** [review]
Tested this pretty heavily along a bunch of different actions and this looks to be 💯 😊

**Comment by janpio** [comment]
Whew... I am relieved ;)
